In [35]:
%load_ext autoreload 
%autoreload 2

import pandas as pd
import analysis_kit as ak

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [36]:
%reload_ext autoreload

In [3]:
import os
os.getcwd()

'C:\\Users\\user\\Documents\\Coursera'

In [15]:
hfi=ak.get_hfi_returns()
hfi.head()

Convertible Arbitrage  CTA Global  Distressed Securities  \
date                                                                
1997-01                   1.19        3.93                   1.78   
1997-02                   1.23        2.98                   1.22   
1997-03                   0.78       -0.21                  -0.12   
1997-04                   0.86       -1.70                   0.30   
1997-05                   1.56       -0.15                   2.33   

         Emerging Markets  Equity Market Neutral  Event Driven  \
date                                                             
1997-01              7.91                   1.89          2.13   
1997-02              5.25                   1.01          0.84   
1997-03             -1.20                   0.16         -0.23   
1997-04              1.19                   1.19         -0.05   
1997-05              3.15                   1.89          3.46   

         Fixed Income Arbitrage  Global Macro  Long/Short Equity  \
date                                                               
1997-01                    1.91          5.73               2.81   
1997-02                    1.22          1.75              -0.06   
1997-03                    1.09         -1.19              -0.84   
1997-04                    1.30          1.72               0.84   
1997-05                    1.18          1.08               3.94   

         Merger Arbitrage  Relative Value  Short Selling  Funds Of Funds  
date                                                                      
1997-01              1.50            1.80          -1.66            3.17  
1997-02              0.34            1.18           4.26            1.06  
1997-03              0.60            0.10           7.78           -0.77  
1997-04             -0.01            1.22          -1.29            0.09  
1997-05              1.97            1.73          -7.37            2.75

#### First, lets go on and see if the distribution of returns among these hedge funds category are distributed normally?

##### One way to do that is to compare the mean and median of each of the category. 

##### If mean < Median, then the distribution is negatively skewed. 

In [16]:
pd.concat([hfi.mean(),hfi.median(),hfi.mean()<hfi.median()], axis="columns") # Columns because we wanna concatenate the 
                                                                            # columns

0     1      2
Convertible Arbitrage   0.550837  0.65   True
CTA Global              0.407376  0.14  False
Distressed Securities   0.694601  0.89   True
Emerging Markets        0.625323  0.96   True
Equity Market Neutral   0.449810  0.51   True
Event Driven            0.634449  0.84   True
Fixed Income Arbitrage  0.436502  0.55   True
Global Macro            0.540304  0.38  False
Long/Short Equity       0.633080  0.79   True
Merger Arbitrage        0.535589  0.60   True
Relative Value          0.579202  0.67   True
Short Selling          -0.170076 -0.53  False
Funds Of Funds          0.426236  0.52   True

 #### As we can see above, 10 out of 13 hedge funds type's return distribution is negatively skewed since the mean returns is less than their median returns. 

Alright, if we now already know that most of these are negatively skewed distribution, is there any way to exactly know how much is the skewness for each of these?

In [18]:
ak.skewness(hfi)

Convertible Arbitrage    -2.639592
CTA Global                0.173699
Distressed Securities    -1.300842
Emerging Markets         -1.167067
Equity Market Neutral    -2.124435
Event Driven             -1.409154
Fixed Income Arbitrage   -3.940320
Global Macro              0.982922
Long/Short Equity        -0.390227
Merger Arbitrage         -1.320083
Relative Value           -1.815470
Short Selling             0.767975
Funds Of Funds           -0.361783
dtype: float64

Dont you think it would be clear if we were to sort these values from largest to smallest skew?

#### Sure

In [20]:
ak.skewness(hfi).sort_values(ascending=True)

Fixed Income Arbitrage   -3.940320
Convertible Arbitrage    -2.639592
Equity Market Neutral    -2.124435
Relative Value           -1.815470
Event Driven             -1.409154
Merger Arbitrage         -1.320083
Distressed Securities    -1.300842
Emerging Markets         -1.167067
Long/Short Equity        -0.390227
Funds Of Funds           -0.361783
CTA Global                0.173699
Short Selling             0.767975
Global Macro              0.982922
dtype: float64

#### To calculate the skewness, we created our own module but you can also do this exact calculation by using a method like scipy.stats.skew(hfi)

So, how much skewness would the normal distribution of returns would have? Is it 0?

#### Yes, it is 0. I can prove that to you by just generating set of random numbers and simulating it

In [22]:
hfi.shape

(263, 13)

In [23]:
import numpy as np
normal_rets=np.random.normal(0, 0.15, size=(263,1))

In [24]:
ak.skewness(normal_rets)

0.12022264898884127

## OR

In [31]:
import scipy.stats
scipy.stats.skew(normal_rets)

array([0.00885168])

#### As you can see, 0.1202 is not that very far from 0. Moreover, if we were to increase the size i.e. the number of observations, the result would be more close to 0

In [25]:
normal_rets=np.random.normal(0, 0.15, size=(26300,1))
ak.skewness(normal_rets)

0.008851680910768043

#### There you go....!

## Now its time for Kurtosis

##### First, we need to go to our module kit and update the formula for kurtosis

In [28]:
ak.kurtosis(hfi).sort_values(ascending=False)

Fixed Income Arbitrage    29.842199
Convertible Arbitrage     23.280834
Equity Market Neutral     17.218555
Relative Value            12.121208
Emerging Markets           9.250788
Merger Arbitrage           8.738950
Event Driven               8.035828
Distressed Securities      7.889983
Funds Of Funds             7.070153
Short Selling              6.117772
Global Macro               5.741679
Long/Short Equity          4.523893
CTA Global                 2.952960
dtype: float64

#### Note that the normally distributed returns have a kurtosis of 3. Anything more than 3 is called excess kurtosis.

#### You might have already noticed that some hedge fund category have kurtosis of around 30. That is LITERALLY 10 times the kurtosis of a normally distributed returns

#### Lets compare the above with the randomly generated normal returns

In [29]:
ak.kurtosis(normal_rets)

3.0154570608328854

## OR

In [33]:
import scipy.stats
scipy.stats.kurtosis(normal_rets) 

array([0.01545706])

#### Now you might be thinking, "What is going on here? I thought the kurtosis for N.D was 3"

#### That is a valid question and the devil is in the detail here.

#### The reason why we are seeing a value near to 0 is because the method .kurtosis(), instead of resulting in kurtosis, is giving us a value of excess kurtosis. 

#### Excess kurtosis= kurtosis-3

#### SO we need to remember this point:

##### Our module that we created results in kurtosis but the scipy method kurtosis results in excess kurtosis

Alright, I got that. 

The next question I have is "How can we say with confidence that the result that we got for skewness and kurtosis are not just due to luck or a random chance?"

##### Thats a great question and the answer to that is we have to use a test called Jarque-Bera test to find out the answer.

In [34]:
# First, lets do the test for normally dist returns

scipy.stats.jarque_bera(normal_rets)

Jarque_beraResult(statistic=0.6052613503575549, pvalue=0.7388719297286692)

##### From the output, particularly the p-value, we can infer that the randomly generated returns are highly significant. 

##### The null hypothesis was "The returns are normally distributed". Since p-value is 0.73, this means no matter which significance level you choose i.e. 1% or 5%, we cannot gether evidence from the data to reject this null hypo. 

#### Hence, we can say the returns data is normally distributed. 

#### Now, lets run the jarque bera test on the hfi returns

##### But before that, we gotta update our analysis kit.

##### And lets first, apply the test to normal returns using our toolkit

In [37]:
ak.is_normal(normal_rets)

True

#### Now, on hedge fund returns

In [38]:
ak.is_normal(hfi)

False

#### We have many categories in hfi. The above result is not what we actually wanted. How can we measure the returns of each fund category and determine whetehr its normal or not?

In [39]:
hfi.aggregate(ak.is_normal)

Convertible Arbitrage     False
CTA Global                 True
Distressed Securities     False
Emerging Markets          False
Equity Market Neutral     False
Event Driven              False
Fixed Income Arbitrage    False
Global Macro              False
Long/Short Equity         False
Merger Arbitrage          False
Relative Value            False
Short Selling             False
Funds Of Funds            False
dtype: bool

##### The aggregate method takes the function that you give it and its gonna apply this function on every column and give you the result

##### Do you remember that we also have the ff data for small and large caps? Lets check their skewness and kurtosis.

In [43]:
ffme=ak.get_ffme_returns()
ak.skewness(ffme)

SmallCap    4.410739
LargeCap    0.233445
dtype: float64

#### Interestingly, the small caps have positive skewness

In [44]:
ak.kurtosis(ffme)

SmallCap    46.845008
LargeCap    10.694654
dtype: float64

##### Now, Jarque bera test on Small and Large cap stocks

In [45]:
ak.is_normal(ffme)

False

#### Again, this is not what we want. We want different output for each column. For that, we have to use the .aggregate() method

In [46]:
ffme.aggregate(ak.is_normal)

SmallCap    False
LargeCap    False
dtype: bool

#### The False here indicates that for both small and large cap stocks, their p-values are less than 0.01. This means they are not significantly normally distributed. 

#### Or in other words, since the p-value is less than 1%, the null hypo that the returns of small and large caps are normally distributed is rejected.